service/history/replication/executable_history_task.go

```diff
@@ -297,3 +302,145 @@ func (e *ExecutableHistoryTask) getDeserializedEvents() (_ []*historypb.HistoryE
+		eventsDesResponse: &eventsDeserializeResponse{
+			events:       append(currentEvents, incomingEvents...),
+			newRunEvents: append(currentNewRunEvents, incomingNewRunEvents...),
+			err:          nil,
+		},
+		wantToBatch: true,
+	}, nil
+}
+
+func (e *ExecutableHistoryTask) validateIncomingBatchTask(incomingTask BatchableTask) (*ExecutableHistoryTask, error) {
+	incomingHistoryTask, isHistoryTask := incomingTask.(*ExecutableHistoryTask)
+	if !isHistoryTask {
+		return nil, serviceerror.NewInvalidArgument("Unsupported Batch type")
+	}
+
+	if err := e.checkWorkflowKey(incomingHistoryTask.WorkflowKey); err != nil {
+		return nil, err
+	}
+
+	if err := e.checkVersionHistoryItem(incomingHistoryTask.versionHistoryItems); err != nil {
+		return nil, err
+	}
+
+	if err := e.checkBaseExecutionInfo(incomingHistoryTask.baseExecutionInfo); err != nil {
+		return nil, err
+	}
+
+	events, newRunEvents, err := incomingHistoryTask.getDeserializedEvents()
+	if err != nil {
+		return nil, err
+	}
+
+	if err = e.checkEvents(events, newRunEvents); err != nil {
+		return nil, err
+	}
+
+	return incomingHistoryTask, nil
+}
+
+// checkVersionHistoryItem will check if incoming tasks Version history is on the same branch as the current one
+func (e *ExecutableHistoryTask) checkVersionHistoryItem(incomingHistoryItems []*historyspb.VersionHistoryItem) error {
+	if len(e.versionHistoryItems) == 0 || len(incomingHistoryItems) == 0 {
+		return serviceerror.NewInvalidArgument("version history empty")
+	}
+
+	firstNotEqualIndex := 0
+	for firstNotEqualIndex < len(e.versionHistoryItems) && firstNotEqualIndex < len(incomingHistoryItems) &&
+		e.versionHistoryItems[firstNotEqualIndex].Equal(incomingHistoryItems[firstNotEqualIndex]) {
+		firstNotEqualIndex++
+	}
+	currentLastIndex := len(e.versionHistoryItems) - 1
+	currentLastItem := e.versionHistoryItems[currentLastIndex]
+	if firstNotEqualIndex == len(e.versionHistoryItems) || // current version history is part of incoming version history
+		(firstNotEqualIndex == currentLastIndex &&
+			currentLastItem.Version == incomingHistoryItems[firstNotEqualIndex].Version &&
+			currentLastItem.EventId <= incomingHistoryItems[firstNotEqualIndex].EventId) {
+		return nil
+	}
+
+	return serviceerror.NewInvalidArgument("version history does not match")
+}
+
+func (e *ExecutableHistoryTask) checkWorkflowKey(incomingWorkflowKey definition.WorkflowKey) error {
+	if !(e.WorkflowKey.GetWorkflowID() == incomingWorkflowKey.GetWorkflowID() &&
+		e.WorkflowKey.GetNamespaceID() == incomingWorkflowKey.GetNamespaceID() &&
+		e.WorkflowKey.GetRunID() == incomingWorkflowKey.GetRunID()) {
+		return serviceerror.NewInvalidArgument("workflow key does not match")
+	}
+	return nil
+}
+
+func (e *ExecutableHistoryTask) checkBaseExecutionInfo(incomingTaskExecutionInfo *workflowpb.BaseExecutionInfo) error {
+	if e.baseExecutionInfo != nil && incomingTaskExecutionInfo != nil && !e.baseExecutionInfo.Equal(incomingTaskExecutionInfo) {
```

Comment:
what about nil case?

Reply:
BaseExecutionInfo is only for a certain usecase and it is either nil or equal. @wxing1292 could you confirm this?

Reply:
should also check if 

```go
if e.baseExecutionInfo == nil && incomingTaskExecutionInfo == nil {
    return nil
} else if e.baseExecutionInfo != nil && incomingTaskExecutionInfo != nil {
    if !e.baseExecutionInfo.Equal(incomingTaskExecutionInfo) {
        return err
    }
    return nil
} else {
    return err
}
````